### ISSUE TITLE
Cannot use both TypeGuard and bool in return type for same `@overload`

### ISSUE THREAD

**Issue by davidfstr** [microsoft/pyright#3215]
**Describe the bug**
It appears that you cannot have a pair of `@overload` cases where one returns a `TypeGuard[T]` and the other returns a `bool`.

**To Reproduce**

* With the following two files, with Python 3.8 installed:

```
# pyproject.toml
[tool.pyright]
include = ["."]
pythonPlatform = "All"

# trycast_mini.py
from typing import get_args, get_origin, Literal, overload, Type, TypeVar, Union
from typing_extensions import TypeGuard

_T = TypeVar("_T")

@overload
def isassignable(value: object, tp: Type[_T]) -> TypeGuard[_T]: ...
@overload
def isassignable(value: object, tp: object) -> bool: ...

def isassignable(value, tp):
    if get_origin(tp) is Union:
        if get_args(tp) == (int, str):
            return isinstance(value, int) or isinstance(value, str)
        raise NotImplementedError()
    return isinstance(value, tp)
```

* ...run the following commands:
    * `pip3 install pyright==1.1.230`
    * `pyright`

**Expected behavior**
There should be no typechecking errors.

**Actual behavior**

Pyright gives the following error:
```
/Users/me/tmp/trycast_mini.py
  /Users/me/tmp/trycast_mini.py:19:5 - error: Overloaded function implementation is not consistent with signature of overload 1
    Function return type "TypeGuard[_T@isassignable]" is incompatible with type "bool"
      "TypeGuard[_T@isassignable]" is incompatible with "bool" (reportGeneralTypeIssues)
1 errors, 0 warnings, 0 informations
```

Mypy gives no error on the same code.

**Additional context**
I am trying to get [isassignable()](https://github.com/davidfstr/trycast#isassignable) from the example above to typecheck in _both_ Pyright and Mypy. Mypy differs from Pyright in that it does **not** consider a _non-class_ type form such as `Union[int, str]` to be compatible with `Type[T]`; it only considers it to be compatible with `object`. Therefore I have to use an `@overload` that has both a `(Type[T]) -> TypeGuard[T]` case and an `(object) -> bool` case in order to have a function that accepts all type forms that works in both typecheckers. Unfortunately Pyright doesn't appear to like both `TypeGuard[T]` and `bool` appearing in the return type for different cases of the same `@overload`.

**Comment by erictraut** [comment]
This works fine if you add the appropriate type annotations to the implementation:
```python
def isassignable(value, tp: Type[_T] | object) -> bool | TypeGuard[_T]:
```

**Comment by davidfstr** [comment]
(1) @erictraut , can you explain the return type of `TypeGuard[_T] | bool`? It feels contradictory to me when I think about it:

* `TypeGuard[_T]` means "this is a special bool value that when true allows a type checker to narrow the type of the first argument (`value`) to be of type `_T` in subsequent code.
* `bool` means "this is a garden variety bool, with no additional implied semantics".
* `TypeGuard[_T] | bool` to me reads as equivalent to `bool`, the least restrictive union member. But that doesn't seem to be what you're intending here.

(2) As a separate question, a callable type that takes `(Type[_T] | object) -> TypeGuard[_T]` - I've intentionally removed`bool` from the return type here - also doesn't make sense to me:

* If the type checker determines that the `object` part of the union matches, it will have no information to define what `_T` is supposed to be bound to.
* If `_T` is unbound, the return type of `TypeGuard[_T]` doesn't seem meaningful to me; I'm not sure what special meaning (if any) `TypeGuard` is supposed to imply in that case.

**Comment by erictraut** [comment]
An overload implementation must return the union of all return types from the individual overloads, hence `bool | TypeGuard[_T]`.

I'll note that this is a very odd use of a type guard function. I would recommend against using it in an overload like this.

And yes, parameters that are annotated with a union that includes a type variable `_T` (or `Type[_T]`) and some other concrete type are problematic because, as you point out, the type variable may go "unsolved". For this reason, I generally try to avoid this usage in my code.

**Comment by davidfstr** [comment]
I did some more experimentation, and was able to preserve the 2 overloads I needed with two tricks:

* Use `cast(TypeGuard[_T], ...)` to cause the `(...) -> TypeGuard[_T]` overload to match.
* Use `if _ALWAYS_FALSE: return _ALWAYS_FALSE` to cause the `(...) -> bool` overload to match.

 However the `TypeGuard[_T]` no longer seems to work, if you examine the line below where `reveal_type(x)  # expect str, actual object 💥`: 

```python
from typing import cast, get_args, get_origin, Literal, overload, Type, TypeVar, Union
from typing_extensions import TypeGuard

_T = TypeVar("_T")

@overload
def isassignable(value: object, tp: Type[_T]) -> TypeGuard[_T]: ...
@overload
def isassignable(value: object, tp: object) -> bool: ...

_ALWAYS_TRUE: bool = True
_ALWAYS_FALSE: bool = False

def isassignable(value, tp):
    # HACK: Trick pyright into thinking this function can return a bool,
    #       to allow the @overload of `(...) -> bool`, which is needed by mypy.
    if not _ALWAYS_TRUE: return _ALWAYS_TRUE
    if _ALWAYS_FALSE: return _ALWAYS_FALSE
    
    if get_origin(tp) is Union:
        if get_args(tp) == (int, str):
            return cast(
                TypeGuard[_T],
                isinstance(value, int) or isinstance(value, str)
            )
        raise NotImplementedError()
    return cast(
        TypeGuard[_T],
        isinstance(value, tp)
    )

x = object()
reveal_type(x)  # object 👍🏻
if isassignable(x, str):
    reveal_type(x)  # expect str, actual object 💥
else:
    reveal_type(x)  # object 👍🏻
```

Any insight as to how I could get the `(...) -> TypeGuard[_T]` overload to work here?

**Comment by erictraut** [comment]
Yeah, you're using `TypeGuard` in a manner in which it was not intended to be used. Like I said, I wouldn't recommend using it in an overload. You should separate the polymorphic function into two distinct functions. One is a type guard, the other is not.

**Comment by davidfstr** [comment]
> I wouldn't recommend using [TypeGuard] in an overload.

That's unfortunate. I will have to find another way to annotate this function.

If [TypeForm[_T]](https://github.com/python/mypy/issues/9773) was a thing, I wouldn't need to use an overload here to preserve compatibility with mypy.

**Comment by davidfstr** [comment]
For future readers, I was able to write an approximate solution for the "I want to write a callable type like ([TypeForm[_T]](https://github.com/python/mypy/issues/9773)) -> TypeGuard[_T]" problem while avoiding the use of TypeGuard[_T] in an overload (which pyright appears to not like) by approximating with:

```
def isassignable(value: object, tp: Type[_T]|object) -> TypeGuard[_T]:
```

This workaround does result in some strange guarding behavior when I pass a non-Type [TypeForm](https://github.com/python/mypy/issues/9773) as an argument:

```python
# trycast_mini.py
from typing import cast, get_args, get_origin, Literal, overload, TYPE_CHECKING, Type, TypeVar, Union
from typing_extensions import TypeGuard

_T = TypeVar("_T")

def isassignable(value: object, tp: Type[_T]|object) -> TypeGuard[_T]:
    if get_origin(tp) is Union:
        if get_args(tp) == (int, str):
            return cast(
                TypeGuard[_T],
                isinstance(value, int) or isinstance(value, str)
            )
        raise NotImplementedError()
    return cast(
        TypeGuard[_T],
        isinstance(value, tp)  # type: ignore[reportGeneralTypeIssues]  # pyright
    )

x = object()
reveal_type(x)  # object 👍🏻
if isassignable(x, str):
    reveal_type(x)  # str 👍🏻
else:
    reveal_type(x)  # object 👍🏻

y = object()
reveal_type(y)  # object 👍🏻
if isassignable(y, Union[str, int]):
    reveal_type(y)  # expect Union[str, int], actual `_T@isassignable` 🤔 
else:
    reveal_type(y)  # object 👍🏻
```

**Comment by davidfstr** [comment]
(For future readers:) Unfortunately, if I run _mypy_ on the above workaround code, I regress my ability to get the correct guarding behavior when matching TypeForms that are Types: (See line below marked with `💥`)

```python
... # def isassignable...

x = object()
reveal_type(x)  # mypy: object 👍🏻
if isassignable(x, str):
    reveal_type(x)  # mypy: expect str, actual `_T`-1` 💥
else:
    reveal_type(x)  # mypy: object 👍🏻

y = object()
reveal_type(y)  # object 👍🏻
if isassignable(y, Union[str, int]):
    reveal_type(y)  # mypy: expect Union[str, int], actual `_T`-1` 🤔 
else:
    reveal_type(y)  # mypy: object 👍🏻
```

So I'm going to have to toss out this latest workaround as well.
